File: template/robot.py
# import library for robot programming
import wpilib
import wpilib.drive
import rev

class MyRobot(wpilib.TimedRobot):
    def robotInit(self):
        # Define what gamepad I'm using
        self.pad = wpilib.XboxController(0)

        # Define drive motors 
        self.motor_left_1 = rev.CANSparkMax(2, rev.CANSparkLowLevel.MotorType.kBrushless)
        self.motor_left_2 = rev.CANSparkMax(3, rev.CANSparkLowLevel.MotorType.kBrushless)
        self.motor_right_1 = rev.CANSparkMax(4, rev.CANSparkLowLevel.MotorType.kBrushless)
        self.motor_right_2 = rev.CANSparkMax(5, rev.CANSparkLowLevel.MotorType.kBrushless)

        # Set the right side to be inverted
        self.motor_right_1.setInverted(True)  # Set right 1 inverted
        self.motor_right_2.setInverted(True)  # SEt right 2 inverted

        # Combine motors into groups
        self.left_group = wpilib.MotorControllerGroup(self.motor_left_1, self.motor_left_2)
        self.right_group = wpilib.MotorControllerGroup(self.motor_right_1, self.motor_right_2)

        # Define drivetrain
        self.drivetrain = wpilib.drive.DifferentialDrive(self.left_group, self.right_group)

    def teleopInit(self):
        pass

    def teleopPeriodic(self):
        # This function gets called over and over again during teleoperated mode

        # 1. Read button / joystick
        left_y = self.pad.getLeftY()
        left_x = self.pad.getLeftX()


        # 2. Calculate 
        speed = left_y
        rotation = -left_x


        # 3. Turn motors
        self.drivetrain.arcadeDrive(speed, rotation)


if __name__ == "__main__":
    wpilib.run(MyRobot)

File: java-template/src/main/java/frc/robot/Robot.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot;

import edu.wpi.first.util.sendable.SendableRegistry;
import edu.wpi.first.wpilibj.TimedRobot;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.XboxController;
import edu.wpi.first.wpilibj.drive.DifferentialDrive;

import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.config.SparkMaxConfig;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.SparkBase.ResetMode;
import com.revrobotics.spark.SparkBase.PersistMode;


/**
 * The methods in this class are called automatically corresponding to each mode, as described in
 * the TimedRobot documentation. If you change the name of this class or the package after creating
 * this project, you must also update the manifest file in the resource directory.
 */
public class Robot extends TimedRobot {
  private final SparkMax m_leftLeader = new SparkMax(2, MotorType.kBrushless);
  private final SparkMax m_leftFollower = new SparkMax(3, MotorType.kBrushless);
  private final SparkMax m_rightLeader = new SparkMax(4, MotorType.kBrushless);
  private final SparkMax m_rightFollower = new SparkMax(5, MotorType.kBrushless);

  private final DifferentialDrive m_robotDrive =
      new DifferentialDrive(m_leftLeader::set, m_rightLeader::set);
  private final XboxController m_controller = new XboxController(0);

  /** Called once at the beginning of the robot program. */
  public Robot() {
    // For dashboards and LiveWindow
    SendableRegistry.addChild(m_robotDrive, m_leftLeader);
    SendableRegistry.addChild(m_robotDrive, m_leftFollower);
    SendableRegistry.addChild(m_robotDrive, m_rightLeader);
    SendableRegistry.addChild(m_robotDrive, m_rightFollower);

    // This is the 2025 REV config structure
    SparkMaxConfig leftLeaderConfig = new SparkMaxConfig();
    leftLeaderConfig
        .idleMode(IdleMode.kBrake);
    m_leftLeader.configure(leftLeaderConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);

    SparkMaxConfig leftFollowerConfig = new SparkMaxConfig();
    leftFollowerConfig
        .idleMode(IdleMode.kBrake)
        .follow(m_leftLeader);
    m_leftFollower.configure(leftFollowerConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);

    SparkMaxConfig rightLeaderConfig = new SparkMaxConfig();
    rightLeaderConfig
        .inverted(true)
        .idleMode(IdleMode.kBrake);
    m_rightLeader.configure(rightLeaderConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);

    SparkMaxConfig rightFollowerConfig = new SparkMaxConfig();
    rightFollowerConfig
        .inverted(true)
        .idleMode(IdleMode.kBrake)
        .follow(m_rightLeader);
    m_rightFollower.configure(rightFollowerConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
  }

  /** This function is run once each time the robot enters autonomous mode. */
  @Override
  public void autonomousInit() {}

  /** This function is called periodically during autonomous. */
  @Override
  public void autonomousPeriodic() {}

  /** This function is called once each time the robot enters teleoperated mode. */
  @Override
  public void teleopInit() {}

  /** This function is called periodically during teleoperated mode. */
  @Override
  public void teleopPeriodic() {

    // 1. Read button / joystick
    double left_y = m_controller.getLeftY();
    double right_x = m_controller.getRightX();

    // 2. Calculate
    double speed = left_y;
    double rotation = -right_x;

    // 3. Turn motors
    m_robotDrive.arcadeDrive(speed, rotation);
  }

  /** This function is called once each time the robot enters test mode. */
  @Override
  public void testInit() {}

  /** This function is called periodically during test mode. */
  @Override
  public void testPeriodic() {}
}
